Allow IP CIDR Range or Any as Source and/or Destination Policy Groups

akgiesler opened the pull request:
Adds a new validator specific to source and destination policy groups.
NSX-T can accept an IP, Range, CIDR, or a Group Path as a source
and/or destination group through the security policy interface.

Updates the getSecurityPolicyAndGatewayRulesSchema function to use the
new validator.

Resolves: Issue #584

annakhm commented:
Thank you @akgiesler! This looks great, but a documentation update is also needed, as well as acc tests. If you can add those it will be great, otherwise we'll follow up with those on top of your change.

akgiesler commented:
Sure thing @annakhm , we are working on those now.

akgiesler commented:
Hi @annakhm, 
I have updated the PR to include tests for Security Policy and Gateway Policy as they share the same function to get & validate their schema.  I have also updated the docs for both to reference the availability of IP, CIDR, or IP ranges.  

During our build-out of the tests, it was discovered that the provider already supports using "ANY" as a source or destination in security policy.  The way to do this was to map the source_groups or destination_groups to an empty set.

for example:
```
  rule {
    display_name       = "rule1"
    source_groups      = [ nsxt_policy_group.group1.path ]
    destination_groups = [ ]
    action             = "ALLOW"
    services           = [data.nsxt_policy_service.icmp.path]
  }
```
This was tested in NSX-T 3.1 using NSX-T Provider v3.1.0.  I added this feature into the Security policy documentation.  I _did not_ add this feature to the gateway policy documentation as I was unable to test it.

annakhm commented on nsxt/resource_nsxt_policy_gateway_policy_test.go:
```diff
@@ -384,6 +384,204 @@ func TestAccResourceNsxtGlobalPolicyGatewayPolicy_withDomain(t *testing.T) {
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testAccNsxtPolicyGatewayPolicyWithIPCidrRangeDestCreate(name, policyIP, policyCidr, policyRange),
```
I would suggest to use a single function to test both Create and Update use cases (just pass different values in params), and also I think one test should be enough to cover both source and dest use cases. While our goal is to have good test coverage, we also try to keep CI running time reasonable :)

akgiesler replied:
I went ahead and collapsed the tests into a single test for both security policy and gateway policy.  I also reduced the terraform generating functions down to single functions as you requested.

annakhm commented on website/docs/r/policy_gateway_policy.html.markdown:
```diff
@@ -93,7 +93,7 @@ The following arguments are supported:
   * `display_name` - (Required) Display name of the resource.
   * `description` - (Optional) Description of the resource.
-  * `destination_groups` - (Optional) A list of destination group paths to use for the policy.
+  * `destination_groups` - (Optional) Set of group paths, IPs, IP ranges, or CIDRs that serve as the destination for this rule.
```
It should be helpful specifying here that IP/ranges/cidrs are only supported starting NSX 3.0.

akgiesler replied:
Hi @annakhm, I have updated all of the documentation as requested.

annakhm commented:
Thanks for those changes @akgiesler! You are correct, "ANY" is equivalent to empty set. I've ran your tests on gateway policy and they pass fine. I noticed that docs for gateway policy are updated in your change.
I've left couple of small comments above, another nit would be to also update documentation in `resource_nsxt_policy_predefined_security_policy` and `resource_nsxt_policy_predefined_gateway_policy`.

annakhm commented:
Thank you @akgiesler, merging this!
